tests/ui/cast.rs

```diff
@@ -61,4 +61,15 @@ fn main() {
     1u32 as usize; // Should not trigger any lint
     1i32 as isize; // Neither should this
     1i32 as usize;
+
+    // Test cast_unnecessary
+    1i32 as i32;
+    1f32 as f32;
+    false as bool;
+    &1i32 as &i32;
+
+    1i32 as i64;      // Should not trigger
```

Comment:
Also, I don't understand your point for `FromStr`.
`i32::from_str("1").unwrap() as i32` is just the same as `1i32 as i32`. Could you explain more?